electrumsv/wallet.py

```diff
@@ -3671,24 +3734,65 @@ def find_account_for_extended_public_key(self, extended_public_key: XPublicKey)
+        # i.e. we may already have received the mAPI callback for the 'correct chain'
+        assert self._db_context is not None
+        assert self.main_headers_client is not None
+        proofs_on_main_chain: List[TransactionProofRow] = []
+        remaining_tx_hashes = set(tx_hashes)
+        tx_proofs_rows = await read_merkle_proofs(self._db_context, tx_hashes)
+        for proof_row in tx_proofs_rows:
+            block_hash, tx_hash, proof_data, block_position = proof_row
+            header, chain = cast(Headers, app_state.headers).lookup(block_hash)
+            if chain == self.main_headers_client.chain:
+                proofs_on_main_chain.append(proof_row)
+                remaining_tx_hashes.remove(tx_hash)
+        return remaining_tx_hashes, tx_proofs_rows
+
+    async def on_reorg(self, orphaned_block_hashes: List[bytes]) -> None:
+        '''Called by network when a reorg has happened'''
+        assert self._db_context is not None
+        assert self.main_headers_client is not None
+        if self._stopped:
+            block_hashes_as_str = [hash_to_hex_str(h) for h in orphaned_block_hashes]
+            self._logger.debug("Cannot undo verifications on a stopped wallet. "
+                "Orphaned block hashes: %s", block_hashes_as_str)
+            return
 
-    #     tx_hashes = db_functions.read_reorged_transactions(self.get_db_context(), above_height)
-    #     self._logger.info('removing verification of %d transactions above %d',
-    #         len(tx_hashes), above_height)
-    #     future = db_functions.set_transactions_reorged(self.get_db_context(), tx_hashes)
-    #     future.result()
+        tx_hashes = await read_reorged_transactions_async(self._db_context, orphaned_block_hashes)
+
+        block_hashes_as_str = [hash_to_hex_str(h) for h in orphaned_block_hashes]
+        self._logger.info('removing verification of %d transactions. "Orphaned block hashes: %s',
+            len(tx_hashes), block_hashes_as_str)
+        await set_transactions_reorged_async(self._db_context, tx_hashes)
+
+        remaining_tx_hashes, proofs_on_main_chain = await self.try_get_mapi_proofs(tx_hashes)
+
+        # Are we expecting a mAPI merkle proof callback for any of these?
+        for_removal = []
+        for tx_hash in remaining_tx_hashes:
+            rows: list[MAPIBroadcastCallbackRow] = self.data.read_mapi_broadcast_callbacks(tx_hash)
+            if len(rows) != 0:
+                for_removal.append(tx_hash)
+        for tx_hash in for_removal:
+            remaining_tx_hashes.remove(tx_hash)
+
+        # Otherwise, register for utxo spend notifications for these transactions to get proof data
+        if len(remaining_tx_hashes) != 0:
+            for tx_hash in remaining_tx_hashes:
+                tx = self.get_transaction(tx_hash)
+                assert tx is not None
+                self._register_spent_outputs_to_monitor(
```

Comment:
Needs a comment about what it means to register these here, and what happens on wallet startup to ensure they are registered when the wallet resumes.